mesos-client/src/main/scala/com/mesosphere/mesos/client/Session.scala

```diff
@@ -0,0 +1,220 @@
+  * @param authorization A [[CredentialsProvider]] if the connection is secured.
+  */
+case class Session(url: URL, streamId: String, authorization: Option[CredentialsProvider] = None) {
+  lazy val isSecured: Boolean = url.getProtocol == "https"
+  lazy val port = if (url.getPort == -1) url.getDefaultPort else url.getPort
+
+  /**
+    * Construct a new [[HttpRequest]] for a serialized Mesos call and a set of authorization, ie session token.
+    * @param bytes The bytes of the serialized Mesos call.
+    * @param maybeCredentials The session token if required.
+    * @return The [[HttpRequest]] with proper headers and body.
+    */
+  def createPostRequest(bytes: Array[Byte], maybeCredentials: Option[HttpCredentials]): HttpRequest = {
+    println(s"Body: ${bytes.map(_.toChar).mkString}")
+    HttpRequest(
+      HttpMethods.POST,
+      uri = Uri(s"$url/api/v1/scheduler"),
+      entity = HttpEntity(MesosClient.ProtobufMediaType, bytes),
+      headers = MesosClient.MesosStreamIdHeader(streamId) :: maybeCredentials.map(Authorization(_)).toList
+    )
+  }
+
+  // Fail and trigger restart if the call was unauthorized.
+  def handleRejection(response: Try[HttpResponse]): Try[HttpResponse] = {
+    response match {
+      case Success(r) if r.status == StatusCodes.Unauthorized =>
+        throw new IllegalStateException("Session token expired.")
+      case id => id
+    }
+  }
+
+  /** @return A flow that transforms serialized Mesos calls to proper HTTP requests. */
+  def post(connection: Flow[HttpRequest, Try[HttpResponse], NotUsed]): Flow[Array[Byte], Try[HttpResponse], NotUsed] =
+    authorization match {
+      case Some(credentialsProvider) =>
+        RestartFlow.withBackoff(1.second, 1.second, 1, 3)(() =>
+          Flow.fromGraph(SessionFlow(credentialsProvider, createPostRequest)).via(connection).map(handleRejection))
+      case None =>
+        Flow[Array[Byte]].map(createPostRequest(_, None)).via(connection)
+    }
+
+  def postSimple(implicit system: ActorSystem): Flow[Array[Byte], Try[HttpResponse], NotUsed] =
+    authorization match {
+      case Some(credentialsProvider) =>
+        implicit val askTimeout = util.Timeout(20.seconds)
+        val sessionActor = system.actorOf(SessionActor.props(credentialsProvider, createPostRequest))
+        Flow[Array[Byte]].ask[HttpResponse](1)(sessionActor).map(Success(_))
+      case None =>
+        Flow[Array[Byte]].map(createPostRequest(_, None)).mapAsync(1)(Http().singleRequest(_)).map(Success(_))
+    }
+
+  /** @return The connection pool for this session. */
+  def connectionPool(implicit system: ActorSystem): Flow[HttpRequest, Try[HttpResponse], NotUsed] = {
+    // Disable pipelining.
+    val poolSettings = ConnectionPoolSettings("").withMaxConnections(1).withPipeliningLimit(1)
+    if (isSecured) {
+      Flow[HttpRequest]
+        .map(_ -> NotUsed)
+        .via(Http().cachedHostConnectionPoolHttps(host = url.getHost, port = port, settings = poolSettings))
```

Comment:
I feel like `cachedHostConnectionPool` is not needed here since one shouldn't cache connections to mesos.

Reply:
We don't need it at all if we opt for the actor solution.